File: backend/app/routers/builds.py
from datetime import date
from decimal import Decimal

from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from app.auth import manager_role_enabled, require_admin, require_auth, require_manager
from app.database import get_db
from app.models import (
    PC,
    BuildComment,
    BuildStatus,
    Part,
    PlannedBuild,
    PlannedBuildItem,
    TransferLog,
)
from app.schemas import (
    BuildCommentIn,
    BuildCommentOut,
    BuildConvertIn,
    BuildCreate,
    BuildDetailOut,
    BuildItemCreate,
    BuildItemOut,
    BuildOut,
    BuildRejectIn,
    BuildUpdate,
    OkOut,
    PCDetailOut,
)
from app.routers.pcs import _get_pc_or_404, _pc_detail

router = APIRouter(prefix="/builds", tags=["planner"])

ADMIN = [Depends(require_admin)]


def _get_build_or_404(db: Session, build_id: str) -> PlannedBuild:
    build = db.execute(
        select(PlannedBuild)
        .options(
            selectinload(PlannedBuild.items)
            .selectinload(PlannedBuildItem.part)
            .selectinload(Part.pc)
        )
        .where(PlannedBuild.id == build_id)
        .execution_options(populate_existing=True)
    ).scalar_one_or_none()
    if build is None:
        raise HTTPException(status_code=404, detail="Build not found")
    return build


def _build_detail(build: PlannedBuild) -> BuildDetailOut:
    out = BuildDetailOut.model_validate(build)
    out.item_count = len(build.items)
    total = Decimal("0")
    for item_out, item in zip(out.items, build.items):
        if item.part is not None:
            if item.part.purchase_price is not None:
                total += item.part.purchase_price
            item_out.part.pc_name = item.part.pc.name if item.part.pc else None
        elif item.external_price is not None:
            total += item.external_price
    out.total_cost = total
    return out


@router.get("", response_model=list[BuildOut])
def list_builds(db: Session = Depends(get_db)):
    builds = (
        db.execute(
            select(PlannedBuild)
            .options(selectinload(PlannedBuild.items))
            .order_by(PlannedBuild.created_at.desc())
        )
        .scalars()
        .all()
    )
    result = []
    for build in builds:
        out = BuildOut.model_validate(build)
        out.item_count = len(build.items)
        result.append(out)
    return result


@router.post("", response_model=BuildDetailOut, status_code=201, dependencies=ADMIN)
def create_build(payload: BuildCreate, db: Session = Depends(get_db)):
    build = PlannedBuild(**payload.model_dump())
    db.add(build)
    db.commit()
    return _build_detail(_get_build_or_404(db, build.id))


@router.get("/{build_id}", response_model=BuildDetailOut)
def get_build(build_id: str, db: Session = Depends(get_db)):
    return _build_detail(_get_build_or_404(db, build_id))


@router.patch("/{build_id}", response_model=BuildDetailOut, dependencies=ADMIN)
def update_build(build_id: str, payload: BuildUpdate, db: Session = Depends(get_db)):
    build = _get_build_or_404(db, build_id)
    for key, value in payload.model_dump(exclude_unset=True).items():
        setattr(build, key, value)
    db.commit()
    return _build_detail(_get_build_or_404(db, build_id))


@router.delete("/{build_id}", response_model=OkOut, dependencies=ADMIN)
def delete_build(build_id: str, db: Session = Depends(get_db)):
    build = _get_build_or_404(db, build_id)
    db.delete(build)
    db.commit()
    return OkOut()


@router.post("/{build_id}/items", response_model=BuildItemOut, status_code=201, dependencies=ADMIN)
def add_build_item(
    build_id: str, payload: BuildItemCreate, db: Session = Depends(get_db)
):
    build = _get_build_or_404(db, build_id)
    if payload.part_id is None and not payload.external_name:
        raise HTTPException(
            status_code=422,
            detail="Provide either part_id (inventory part) or external_name",
        )
    if payload.part_id is not None:
        part = db.get(Part, payload.part_id)
        if part is None:
            raise HTTPException(status_code=404, detail="Part not found")
        already = {i.part_id for i in build.items if i.part_id}
        if payload.part_id in already:
            raise HTTPException(status_code=409, detail="Part already in this build")
    item = PlannedBuildItem(build_id=build.id, **payload.model_dump())
    db.add(item)
    # content changed — any prior manager sign-off no longer applies
    build.status = BuildStatus.draft
    db.commit()
    fresh = db.execute(
        select(PlannedBuildItem)
        .options(selectinload(PlannedBuildItem.part).selectinload(Part.pc))
        .where(PlannedBuildItem.id == item.id)
    ).scalar_one()
    out = BuildItemOut.model_validate(fresh)
    if fresh.part is not None:
        out.part.pc_name = fresh.part.pc.name if fresh.part.pc else None
    return out


@router.delete("/{build_id}/items/{item_id}", response_model=OkOut, dependencies=ADMIN)
def remove_build_item(build_id: str, item_id: str, db: Session = Depends(get_db)):
    item = db.execute(
        select(PlannedBuildItem).where(
            PlannedBuildItem.id == item_id, PlannedBuildItem.build_id == build_id
        )
    ).scalar_one_or_none()
    if item is None:
        raise HTTPException(status_code=404, detail="Build item not found")
    # content changed — any prior manager sign-off no longer applies
    item.build.status = BuildStatus.draft
    db.delete(item)
    db.commit()
    return OkOut()


@router.post("/{build_id}/convert", response_model=PCDetailOut, status_code=201, dependencies=ADMIN)
def convert_build(
    build_id: str, payload: BuildConvertIn, db: Session = Depends(get_db)
):
    """Turn a planned build into a real PC: creates the PC, moves every
    attached inventory part onto it (with transfer logs), then deletes the
    plan. External (not-yet-owned) items are dropped — buy them first.

    When a manager account exists, the build must be approved first."""
    build = _get_build_or_404(db, build_id)
    if manager_role_enabled() and build.status != BuildStatus.approved:
        raise HTTPException(
            status_code=409,
            detail="Build needs manager approval before it can become a real PC "
            f"(current status: {build.status.value}).",
        )
    pc = PC(
        name=payload.name or build.name,
        description=payload.description or build.notes,
        build_date=date.today(),
    )
    db.add(pc)
    db.flush()
    for item in build.items:
        if item.part is not None:
            db.add(
                TransferLog(
                    part_id=item.part.id,
                    from_pc_id=item.part.pc_id,
                    to_pc_id=pc.id,
                )
            )
            item.part.pc_id = pc.id
    db.delete(build)
    db.commit()
    return _pc_detail(_get_pc_or_404(db, pc.id))


# ── approval workflow ────────────────────────────────────────────────────────

@router.post("/{build_id}/submit", response_model=BuildDetailOut, dependencies=ADMIN)
def submit_build(build_id: str, db: Session = Depends(get_db)):
    """Admin sends the build to the manager for review."""
    build = _get_build_or_404(db, build_id)
    if not manager_role_enabled():
        raise HTTPException(status_code=400, detail="No manager account is configured")
    if build.status == BuildStatus.pending:
        raise HTTPException(status_code=409, detail="Already waiting for review")
    if build.status == BuildStatus.approved:
        raise HTTPException(status_code=409, detail="Already approved")
    if len(build.items) == 0:
        raise HTTPException(status_code=422, detail="Add parts before submitting")
    build.status = BuildStatus.pending
    db.commit()
    return _build_detail(_get_build_or_404(db, build_id))


@router.post("/{build_id}/approve", response_model=BuildDetailOut)
def approve_build(
    build_id: str,
    db: Session = Depends(get_db),
    _role: str = Depends(require_manager),
):
    build = _get_build_or_404(db, build_id)
    if build.status != BuildStatus.pending:
        raise HTTPException(
            status_code=409, detail=f"Only pending builds can be approved (status: {build.status.value})"
        )
    build.status = BuildStatus.approved
    db.commit()
    return _build_detail(_get_build_or_404(db, build_id))


@router.post("/{build_id}/reject", response_model=BuildDetailOut)
def reject_build(
    build_id: str,
    payload: BuildRejectIn,
    db: Session = Depends(get_db),
    _role: str = Depends(require_manager),
):
    build = _get_build_or_404(db, build_id)
    if build.status != BuildStatus.pending:
        raise HTTPException(
            status_code=409, detail=f"Only pending builds can be rejected (status: {build.status.value})"
        )
    build.status = BuildStatus.rejected
    if payload.comment and payload.comment.strip():
        db.add(BuildComment(build_id=build.id, author_role="manager", body=payload.comment.strip()))
    db.commit()
    return _build_detail(_get_build_or_404(db, build_id))


# ── discussion thread ────────────────────────────────────────────────────────

@router.get("/{build_id}/comments", response_model=list[BuildCommentOut])
def list_comments(build_id: str, db: Session = Depends(get_db)):
    _get_build_or_404(db, build_id)
    comments = (
        db.execute(
            select(BuildComment)
            .where(BuildComment.build_id == build_id)
            .order_by(BuildComment.created_at.asc())
        )
        .scalars()
        .all()
    )
    return comments


@router.post("/{build_id}/comments", response_model=BuildCommentOut, status_code=201)
def add_comment(
    build_id: str,
    payload: BuildCommentIn,
    request: Request,
    db: Session = Depends(get_db),
):
    role = require_auth(request)  # both roles can join the discussion
    _get_build_or_404(db, build_id)
    comment = BuildComment(build_id=build_id, author_role=role, body=payload.body.strip())
    db.add(comment)
    db.commit()
    return comment

File: backend/routes/builds.js
import { Router } from 'express';
import { PlannedBuild, PlannedBuildItem, BuildComment, PC, Part, TransferLog } from '../config/models.js';
import { requireAdmin, requireManager, requireAuth, managerRoleEnabled } from '../middleware/auth.js';

const router = Router();

async function enrichBuild(build) {
  const items = await PlannedBuildItem.find({ build_id: build._id.toString() }).lean();
  let total = 0;
  const enrichedItems = await Promise.all(items.map(async item => {
    let part = null;
    if (item.part_id) {
      part = await Part.findById(item.part_id).lean();
      if (part) {
        part = { ...part, id: part._id, pc_name: null };
        if (part.pc_id) { const pc = await PC.findById(part.pc_id).lean(); if (pc) part.pc_name = pc.name; }
        total += part.purchase_price || 0;
      }
    } else { total += item.external_price || 0; }
    return { ...item, id: item._id, part };
  }));
  return { ...build, id: build._id, items: enrichedItems, item_count: items.length, total_cost: total };
}

router.get('/', async (req, res, next) => {
  try {
    let builds = await PlannedBuild.find({}).lean();
    builds.sort((a, b) => new Date(b.created_at) - new Date(a.created_at));
    const result = await Promise.all(builds.map(async b => {
      const cnt = await PlannedBuildItem.countDocuments({ build_id: b._id.toString() });
      return { ...b, id: b._id, item_count: cnt };
    }));
    res.json(result);
  } catch (err) { next(err); }
});

router.post('/', requireAdmin, async (req, res, next) => {
  try {
    const build = await PlannedBuild.create({ ...req.body, status: 'draft', created_at: new Date() });
    res.status(201).json(await enrichBuild(build.toObject()));
  } catch (err) { next(err); }
});

router.get('/:buildId', async (req, res, next) => {
  try {
    const build = await PlannedBuild.findById(req.params.buildId).lean();
    if (!build) return res.status(404).json({ detail: 'Build not found' });
    res.json(await enrichBuild(build));
  } catch (err) { next(err); }
});

router.patch('/:buildId', requireAdmin, async (req, res, next) => {
  try {
    const build = await PlannedBuild.findById(req.params.buildId).lean();
    if (!build) return res.status(404).json({ detail: 'Build not found' });
    await PlannedBuild.findByIdAndUpdate(req.params.buildId, { $set: req.body });
    const updated = await PlannedBuild.findById(req.params.buildId).lean();
    res.json(await enrichBuild(updated));
  } catch (err) { next(err); }
});

router.delete('/:buildId', requireAdmin, async (req, res, next) => {
  try {
    const build = await PlannedBuild.findById(req.params.buildId).lean();
    if (!build) return res.status(404).json({ detail: 'Build not found' });
    await Promise.all([
      PlannedBuildItem.deleteMany({ build_id: build._id.toString() }),
      BuildComment.deleteMany({ build_id: build._id.toString() }),
      PlannedBuild.findByIdAndDelete(req.params.buildId),
    ]);
    res.json({ ok: true });
  } catch (err) { next(err); }
});

router.post('/:buildId/items', requireAdmin, async (req, res, next) => {
  try {
    const build = await PlannedBuild.findById(req.params.buildId).lean();
    if (!build) return res.status(404).json({ detail: 'Build not found' });
    const { part_id, external_name } = req.body;
    if (!part_id && !external_name) return res.status(422).json({ detail: 'Provide either part_id (inventory part) or external_name' });
    if (part_id) {
      const part = await Part.findById(part_id).lean();
      if (!part) return res.status(404).json({ detail: 'Part not found' });
      const existing = await PlannedBuildItem.findOne({ build_id: build._id.toString(), part_id }).lean();
      if (existing) return res.status(409).json({ detail: 'Part already in this build' });
    }
    const item = await PlannedBuildItem.create({ build_id: build._id.toString(), ...req.body });
    await PlannedBuild.findByIdAndUpdate(req.params.buildId, { $set: { status: 'draft' } });
    const doc = item.toObject();
    let part = null;
    if (doc.part_id) {
      part = await Part.findById(doc.part_id).lean();
      if (part) { part = { ...part, id: part._id, pc_name: null }; if (part.pc_id) { const pc = await PC.findById(part.pc_id).lean(); if (pc) part.pc_name = pc.name; } }
    }
    res.status(201).json({ ...doc, id: doc._id, part });
  } catch (err) { next(err); }
});

router.delete('/:buildId/items/:itemId', requireAdmin, async (req, res, next) => {
  try {
    const item = await PlannedBuildItem.findOne({ _id: req.params.itemId, build_id: req.params.buildId }).lean();
    if (!item) return res.status(404).json({ detail: 'Build item not found' });
    await PlannedBuild.findByIdAndUpdate(req.params.buildId, { $set: { status: 'draft' } });
    await PlannedBuildItem.findByIdAndDelete(req.params.itemId);
    res.json({ ok: true });
  } catch (err) { next(err); }
});

router.post('/:buildId/convert', requireAdmin, async (req, res, next) => {
  try {
    const build = await PlannedBuild.findById(req.params.buildId).lean();
    if (!build) return res.status(404).json({ detail: 'Build not found' });
    if (await managerRoleEnabled() && build.status !== 'approved')
      return res.status(409).json({ detail: `Build needs manager approval before it can become a real PC (current status: ${build.status}).` });
    const { name, description } = req.body;
    const pc = await PC.create({ name: name || build.name, description: description || build.notes, status: 'active', qr_code: crypto.randomUUID(), build_date: new Date(), created_at: new Date() });
    const pcId = pc._id.toString();
    const items = await PlannedBuildItem.find({ build_id: build._id.toString() }).lean();
    for (const item of items) {
      if (item.part_id) {
        await TransferLog.create({ part_id: item.part_id, from_pc_id: null, to_pc_id: pcId, moved_at: new Date() });
        await Part.findByIdAndUpdate(item.part_id, { $set: { pc_id: pcId } });
      }
    }
    await Promise.all([
      PlannedBuildItem.deleteMany({ build_id: build._id.toString() }),
      BuildComment.deleteMany({ build_id: build._id.toString() }),
      PlannedBuild.findByIdAndDelete(req.params.buildId),
    ]);
    const parts = await Part.find({ pc_id: pcId }).lean();
    res.status(201).json({ ...pc.toObject(), id: pcId, parts: parts.map(p => ({ ...p, id: p._id, pc_name: pc.name })), part_count: parts.length, total_value: parts.reduce((s, p) => s + (p.purchase_price || 0), 0), has_network_info: false });
  } catch (err) { next(err); }
});

// ── approval workflow ────────────────────────────────────────────────────

router.post('/:buildId/submit', requireAdmin, async (req, res, next) => {
  try {
    const build = await PlannedBuild.findById(req.params.buildId).lean();
    if (!build) return res.status(404).json({ detail: 'Build not found' });
    if (!(await managerRoleEnabled())) return res.status(400).json({ detail: 'No manager account is configured' });
    if (build.status === 'pending') return res.status(409).json({ detail: 'Already waiting for review' });
    if (build.status === 'approved') return res.status(409).json({ detail: 'Already approved' });
    const cnt = await PlannedBuildItem.countDocuments({ build_id: build._id.toString() });
    if (cnt === 0) return res.status(422).json({ detail: 'Add parts before submitting' });
    await PlannedBuild.findByIdAndUpdate(req.params.buildId, { $set: { status: 'pending' } });
    const updated = await PlannedBuild.findById(req.params.buildId).lean();
    res.json(await enrichBuild(updated));
  } catch (err) { next(err); }
});

router.post('/:buildId/approve', requireManager, async (req, res, next) => {
  try {
    const build = await PlannedBuild.findById(req.params.buildId).lean();
    if (!build) return res.status(404).json({ detail: 'Build not found' });
    if (build.status !== 'pending') return res.status(409).json({ detail: `Only pending builds can be approved (status: ${build.status})` });
    await PlannedBuild.findByIdAndUpdate(req.params.buildId, { $set: { status: 'approved' } });
    const updated = await PlannedBuild.findById(req.params.buildId).lean();
    res.json(await enrichBuild(updated));
  } catch (err) { next(err); }
});

router.post('/:buildId/reject', requireManager, async (req, res, next) => {
  try {
    const build = await PlannedBuild.findById(req.params.buildId).lean();
    if (!build) return res.status(404).json({ detail: 'Build not found' });
    if (build.status !== 'pending') return res.status(409).json({ detail: `Only pending builds can be rejected (status: ${build.status})` });
    await PlannedBuild.findByIdAndUpdate(req.params.buildId, { $set: { status: 'rejected' } });
    const comment = req.body?.comment;
    if (comment && comment.trim()) await BuildComment.create({ build_id: build._id.toString(), author_role: 'manager', body: comment.trim(), created_at: new Date() });
    const updated = await PlannedBuild.findById(req.params.buildId).lean();
    res.json(await enrichBuild(updated));
  } catch (err) { next(err); }
});

// ── discussion thread ─────────────────────────────────────────────────────

router.get('/:buildId/comments', async (req, res, next) => {
  try {
    const build = await PlannedBuild.findById(req.params.buildId).lean();
    if (!build) return res.status(404).json({ detail: 'Build not found' });
    let comments = await BuildComment.find({ build_id: req.params.buildId }).lean();
    comments.sort((a, b) => new Date(a.created_at) - new Date(b.created_at));
    res.json(comments.map(c => ({ ...c, id: c._id })));
  } catch (err) { next(err); }
});

router.post('/:buildId/comments', requireAuth, async (req, res, next) => {
  try {
    const build = await PlannedBuild.findById(req.params.buildId).lean();
    if (!build) return res.status(404).json({ detail: 'Build not found' });
    const { body } = req.body;
    if (!body?.trim()) return res.status(422).json({ detail: 'Comment body is required' });
    const comment = await BuildComment.create({ build_id: req.params.buildId, author_role: req.userRole, body: body.trim(), created_at: new Date() });
    const doc = comment.toObject();
    res.status(201).json({ ...doc, id: doc._id });
  } catch (err) { next(err); }
});

export default router;
